Use Android KeyStore to encrypt application password credentials

adalpari opened the pull request:
## Description
This PR is adding an extra encryption layer to the username and password of the Application Password credentials.
Credentials are encryopted and encoder into Base64 because of some [problems with WellSQL and Blobs](https://a8c.slack.com/archives/C02QANACA/p1747844050375369)
Each encrypted string has its own IV vector because of security requirements. Otherwise, reusing the same vector for different strings threw an exception.

NOTE: There are some debug lines in the code that will be removed after the review. I added a couple of TODOs to be sure I don't forget about them. (With a TODO, the CI fails)

## Testing instructions

1. Install trunk
2. Log in with a self-hosted site
3. Install this branch to force a DB update
4. In MySite log in with Application Password
- [ ] Verify you see the logs about the storing and the original credentials are the same as the ones after decrypting them
```
WP_RS   com.jetpack.android.beta   D  Original : USERNAME - PASSWORD
WP_RS   com.jetpack.android.beta   D  Encrypted: 5CxycRJ/wefrwergwerfdVkXPPM - krSV2eg+6oVN22AGewgrfwergfwergfeww28/EO3xT
WP_RS   com.jetpack.android.beta   D  Decrypted: USERNAME - PASSWORD
```

5. Kill the app and run it again
- [ ] Verify you see the encrypted and decrypted strings again, and the decrypted ones are correct
```
WP_RS   com.jetpack.android.beta   D  Encrypted: 5CxycRJ/wefrwergwerfdVkXPPM - krSV2eg+6oVN22AGewgrfwergfwergfeww28/EO3xT
WP_RS   com.jetpack.android.beta   D  Decrypted: USERNAME - PASSWORD
```

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

adalpari committed:
Adding encryption utils

adalpari committed:
Changing a bit the helper

adalpari committed:
Returning the IV to store it

adalpari committed:
Storing password as byte aray

adalpari committed:
test

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Using Base64 encoding

adalpari committed:
Encryoting username and password in a row

adalpari committed:
Using two IV, since it's a mandatory practice for encruption safety

adalpari committed:
detekt

adalpari committed:
DB update

adalpari committed:
Merge remote-tracking branch 'origin/trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Fixing typo

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Reusing the same key from the store

adalpari committed:
Merge branch 'feat/CMM-390-use-Android-Keystore-to-encrypt-application-password' of https://github.com/wordpress-mobile/WordPress-Android into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Fixing test

adalpari committed:
Removing tests until findind a solution for it

Encryption test requires Robolectric, but the module is using fux, which is using an old Apache legacy library and it's not supported. Supporting that library in the WP module doesn't make sense, and setting Robolectric API to the legacy one is causing other issues.

useLibrary 'org.apache.http.legacy'
@Config(sdk = [Build.VERSION_CODES.LOLLIPOP_MR1]) // SDK 22, before Apache HTTP was removed

adalpari committed:
detekt

adalpari committed:
Merge branch 'feat/CMM-390-use-Android-Keystore-to-encrypt-application-password' of https://github.com/wordpress-mobile/WordPress-Android into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Fixing test

adalpari committed:
Merge branch 'feat/CMM-390-use-Android-Keystore-to-encrypt-application-password' of https://github.com/wordpress-mobile/WordPress-Android into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/encryption/EncryptionUtils.kt:
```diff
@@ -0,0 +1,88 @@
+private const val CIPHER_TRANSFORMATION_TYPE = "AES/GCM/NoPadding"
+private const val T_LENGTH = 128
+
+@Singleton
```
Do we need this to be a singleton?

adalpari replied:
This class is generating/holding the Secret Key stored in the Android system, so I just wanted to be sure we are not handling it in different parts of the app.
Not a case where we 100% need a singleton, but it looks a bit safer imo.

oguzkocer replied:
In general it'd be best to avoid singletons when they are not strictly necessary. When they are necessary, we should be trying to keep its exposure surface as small as possible. This is because in a multi-threaded environment, it'll create a bottleneck between threads.

In this case, if we need a singleton for generating/holding the key (not sure if we do or not) it'd be better for us to have a type that only does that, but nothing else. Then this new type can be injected into the `EncryptionUtils` without making `EncryptionUtils` a singleton. This might seem like a small difference, but depending on the situation it can make a big difference, because multiple threads can handle encryption/decryption without blocking each other.

In any case, since this class doesn't need to be efficient, I am fine to leave it as is. Whenever I review a PR, I tend to leave these types of comments where the improvement is optional, because if I were to implement this PR I'd likely include it. 🤷‍♂️

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/encryption/EncryptionUtils.kt:
```diff
@@ -0,0 +1,88 @@
+                load(null)
+            }
+            ks.getKey(KEY_STORE_ALIAS, "".toCharArray())
+        } catch (ignore: Throwable) {
+            null
+        }
```
What is the expected behavior when we get an exception here?

It looks to me like we'll generate a new key after each exception and if I am not mistaken, that means the previously encrypted data will not be accessible to us. Assuming that's correct, I think this can lead to a frustrating experience for some of the users because we won't be aware of it unless the user reports it to us as a support request.

We have a few options:
1. Keep the current behavior, but log the exception, so it'll be visible during support requests, crash reports etc
2. Keep the current behavior, but [send a report to Sentry](https://github.com/Automattic/Automattic-Tracks-Android/blob/trunk/crashlogging/src/main/java/com/automattic/android/tracks/crashlogging/CrashLogging.kt#L36-L52)
3. Change the behavior to crash on unexpected exceptions

In general, I prefer the fail fast approach in `3`, but I am not sure if it's the right choice for this flow or not, because I am not sure what our planned final flow looks like. I'll chat with @jkmassel to get a better handle on that.

adalpari replied:
Actually, having an exception means that the Key is not present in the system, so we are then, returning null and forcing the generation of a new one. However, your point is right when saying we can be hiding other exceptions.

Let me see if we can detect when the key is not present, so we are not returning an exception in that case, but crashing the app in other scenarios

adalpari replied:
Ok, I misread the docs, so in case the key is not present in the system, no exception is thrown, and we are getting just null.
So, I've removed the try-catch block altogether, and we should be good to go now with exception reports.

PS: I've also removed all the testing code since you have already tested the PR

oguzkocer reviewed (commented):
Looks good. I left just a couple questions.

I've followed the testing instructions and it worked as expected. However, I found it confusing that the user has to add the site first and then have to go through the authorize application password flow. I think it makes sense to do that if the user is already logged in, but I've tested it as a new user as well and I still had to go through both flows. I assume this is temporary while we implement the whole flow 🤞

adalpari commented:
> I found it confusing that the user has to add the site first and then have to go through the authorization application password flow. I think it makes sense to do that if the user is already logged in, but I've tested it as a new user as well, and I still had to go through both flows. I assume this is temporary while we implement the whole flow 🤞

Sorry for the confusion about that. I created the PR thinking that the reviewer (Nick) was already familiar with the flow. 
The idea behind the login card is to show the Application Password login flow to already logged-in users. This is, to the current users. (But for testing purposes, one has to do both steps 😅 )
The next step will be to allow new users to directly log into the app by using Application Password flow.

adalpari committed:
No need to catch the exception

adalpari committed:
Removing debug code

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password

adalpari committed:
Merge branch 'trunk' into feat/CMM-390-use-Android-Keystore-to-encrypt-application-password
